src/android/com/totalpave/cordova/sqlite/SQLite.java

```diff
@@ -148,6 +151,74 @@ else if (action.equals("close")) {
             callback.success();
             return true;
         }
+        else if (action.equals("backup")) {
+            String path = args.getString(0);
+            String backupPath = args.getString(1);
+
+            File source = $parsePath(path);
+            File destination = $parsePath(backupPath);
+            destination.delete(); // returns false if fails but we don't care.
+
+            FileInputStream in;
+            FileOutputStream out;
+
+            try {
+                in = new FileInputStream(source);
+                out = new FileOutputStream(destination);
+            }
+            catch (FileNotFoundException ex) {
+                callback.error(new SqliteException(Error.DOMAIN, ex.getMessage(), Error.IO_ERROR).toDictionary());
+                return true;
+            }
+            try {
+                // Transfer bytes from in to out
+                byte[] buf = new byte[1024];
```

Comment:
We can optimise this to use a single page of memory.

```java
/**
     * Gets the ideal buffer size for processing streams of data.
     *
     * @return The page size of the device.
     */
    private int getPageSize() {
        // Get the page size of the device. Most devices will be 4096 (4kb)
        // Newer devices may be 16kb
        long ps = Os.sysconf(OsConstants._SC_PAGE_SIZE);

        // sysconf returns a long because it's a general purpose API
        // the expected value of a page size should not exceed an int,
        // but we guard it here to avoid integer overflow just in case
        if (ps > Integer.MAX_VALUE) {
            ps = Integer.MAX_VALUE;
        }

        return (int) ps;
    }
```

Also requires

```java
import android.system.Os;
import android.system.OsConstants;
```

Using `1024` for a buffer is 1/4 of a page, so the OS may allocate 4kb memory, just to use 1kb for the buffer. So at minimum buffers should be 4kb, but Android 15 introduces 16kb page size devices, so the function dynamically fetches the device page size. It will return 4096 or whatever 16kb works out to be in bytes. Then we can utilise an entire page for the buffer which will speed up copyng.

We can also multiply the page size by number of pages to allocate more memory to make copying faster within reason:

```java
int bufferSize = getPageSize() * 2; // 2 pages of memory
byte[] buffer = new byte[bufferSize];
```

Reply:
* 2 pages of memory kinda defeats the purpose of the int max value guard.

Reply:
// sysconf returns a long because it's a general purpose API
        // the expected value of a page size should not exceed an int,
        // but we guard it here to avoid integer overflow just in case
        if (ps > Integer.MAX_VALUE) {
            ps = Integer.MAX_VALUE;
        }

Reply:
You're telling me "we should guard against integer overflow by ensuring the return value is no bigger than max int" and then *2 the return value in the usage. If the guard is necessary then it's theoretically unsafe to *2 the return value. It would cause integer overflow.

Reply:
It's safe to assume that the page size is 4kb or 16kb.

That's there just to appease lints since it will warn if you assign `long` to an `int` and we down size to an `int` so that the value can be used in other APIs.